src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1195,126 +1004,59 @@ public int GetFixedLength()
             return -1;
         }
 
-#if DEBUG
-        private TransitionRegex<TSet>? _transitionRegex;
         /// <summary>
-        /// Computes the symbolic derivative as a transition regex.
-        /// Transitions are in the tree left to right in the order the backtracking engine would explore them.
+        /// Insert <see cref="SymbolicRegexNodeKind.FixedLengthMarker"/> nodes to mark paths in the regex that correspond
+        /// to matches of fixed length. For example, for abar|bar two markers would be added abar(4)|bar(3).
         /// </summary>
-        internal TransitionRegex<TSet> CreateDerivative()
+        /// <remarks>
+        /// This function will rebuild concatenations because it pushes the FixedLengthMarker into the rightmost element.
+        /// Due to this this function should not be called on every character.
+        /// </remarks>
+        /// <param name="lengthSoFar">accumulater used in the recursion for lengths of paths</param>
+        /// <returns>the node with fixed length markers added</returns>
+        public SymbolicRegexNode<TSet> AddFixedLengthMarkers(int lengthSoFar = 0)
         {
-            if (_transitionRegex is not null)
-            {
-                return _transitionRegex;
-            }
-
-            if (IsNothing || IsEpsilon)
-            {
-                _transitionRegex = TransitionRegex<TSet>.Leaf(_builder._nothing);
-                return _transitionRegex;
-            }
-
-            if (IsAnyStar || IsAnyPlus)
-            {
-                _transitionRegex = TransitionRegex<TSet>.Leaf(_builder._anyStar);
-                return _transitionRegex;
-            }
-
             if (!StackHelper.TryEnsureSufficientExecutionStack())
             {
-                return StackHelper.CallOnEmptyStack(CreateDerivative);
+                return StackHelper.CallOnEmptyStack(AddFixedLengthMarkers, lengthSoFar);
```

Comment:
It may not actually matter since we hopefully never hit this path, but I'm wondering if we should just make this return rather than forking to a new thread.  When the correctness of the algorithm depends on continuing, we need to fork like this, but this path is just an optimization, right?  In which case we could just stop optimizing if the expression is so complex that you end up being too deep?

Reply:
Right, that's a good idea. We already did that in `GetFixedLength`, so I'll change it here too.